src/content/articles/for-in-string-types.mdx

```diff
@@ -0,0 +1,212 @@
+---
+date: "February 25 2023"
+description: "Why TypeScript reports a `string` type instead of `keyof typeof` for some forms of loops over objects."
+meta: keyof, loops, objects, strings
+---
+
+# Why `for...in` and `Object.keys()` Give `string` Types
+
+Why does TypeScript report a `string` type instead of `keyof typeof` the object when you use a `for...in` loop or `Object.keys()` on an object?
+
+One of the most commonly asked questions in TypeScript help forums is why code like the following gives an error under the [`noImplicitAny` compiler option](https://www.typescriptlang.org/tsconfig#noImplicitAny).
+
+```ts twoslash
+// @errors: 7053
+const counts = {
+  one: 1,
+  two: 2,
+};
+
+for (const i in counts) {
+  //       ^?
+  console.log(counts[i]);
+}
+```
+
+Shouldn't TypeScript know that `i` is type `"one" | "two"`, a.k.a. `keyof typeof counts`?
+Why is it telling us that `i` is `string` and can't be used like `counts[i]`?
+
+The answer is logical -though often inconvenient- and requires understanding a key characteristic of TypeScript's type system.
+
+<!--truncate-->
+
+## Objects May Have More Properties Than They Appear
+
+TypeScript's type system is _structurally typed_: meaning any value that happens to satisfy a type is allowed to be used as a value of that type.
+In other words, when you declare that a value is of a particular object type, you're telling TypeScript any object that just so happens to have those properties is allowed to be considered that type.
+
+As a result, when you write a `for...in` loop over the keys of an object, you often won't have a guarantee that the object's keys will only be from the type of the object.
+Some other area of code might have provided an object with more properties that just so happens to match the structural type.
```

Comment:
[Proofreading] Since this sentence is describing other parts of the code, consider turning "area" into a plural noun to convey this idea clearly. 
```suggestion
Some other areas of code might have provided an object with more properties that just so happens to match the structural type.
```